### DISCUSSION TITLE
Support for Bosch MH210 Hybrid Manager?

### DISCUSSION THREAD

**Discussion by Domoberry** [emsesp/EMS-ESP32#2138] [Q&A] [answered]
Hi Team,
Having recently upgraded to an E32V2 gtwy (v3.5.6) to link my new Bosch CS5800iAW heat pump with MH210 hybrid manager (part of the CSH5800i box), CT200B and existing 9000i boiler to Domoticz. Again quite amazing what this sw package does!
I understood the MH210 is a relatively new kid on the block (like me in this Q&A section). Which could explain why it is detected as 2  'HM200' devices. EMS-ESP 3.5.6 detects 0 entities on the first HM200 (Device ID 0x20, Product ID 248) and 34 entities on the 2nd HM200 (Device ID 0x53, Product ID 248). For the CT200 no useable entities are detected, I understood this is due to the nature of the Easycontrol CT200.
It would be great if someone could provide some additional background on how (dis)similar MH210 is compared to MH200 and if dedicated focus on the MH210 is planned/needed.
In any case, I remain impressed with the functionality of this sw package - thanks!

**Comment by MichaelDvP** [comment]
I can not find any info about MH210 in web, seems to be very new. 
Product-ID 248 indicates same hardware as HM200, but device-ID 0x20 indicates a mixer function.
Please update to the latest 3.7.0dev, let it run >30 min, then download from help page the system info and attach it here. 
Got to Status->System Log, select loglevel ALL, wait 3 minutes and download the log, also attach it.

**Comment by Domoberry** [comment]
Thanks for your reply!
The MH210 hybrid manager is part of the CSH5800i box, the latter is basically an enclosure for the hybrid manager and a circulation pump. I found a reference to 'MH210' in the Bosch installation manual of the CSH5800i box.
I have attached both files.


On a sidenote: in my case the update 3.6.5 -> 3.7.0dev triggered issues on Domoticz side, for example 
only a smaller subset of the entities make it as Domoticz-devices, many other result in error messages like:
`2024-10-25 14:21:29.844 Error: HybridHP: Exception (GetValueFromTemplate): in Json::Value::resolveReference(key, end): requires objectValue! (Template: value_json[syspress] if value_json[syspress] is defined else 0)`
I need to digest those still. The related mqtt-ad config message and the related mqtt data topic look ok; I'm probably overlooking something.
[241025 emsesp_system_info.json](https://github.com/user-attachments/files/17523399/241025.emsesp_system_info.json)
[241025 log.txt](https://github.com/user-attachments/files/17523401/241025.log.txt)

**Reply by proddy** [reply]
on Domoticz: those conditional templates are used in the MQTT Discovery availability attribute for each Sensor, like:
```json
  "avty": [
    {
      "t": "ems-esp/boiler_data",
      "val_tpl": "{{'online' if value_json['syspress'] is defined else 'offline'}}"
    }
  ]
```

I'm not a Domoticz user so can't validate if its supported. It'll be great if you could help us out. It's easier to adjust in the code.

**Reply by MichaelDvP** [reply]
Thanks, indeed the MH210 publishes mixer telegrams:
```
heatpump(0x20) -W-> boiler(0x08), UBASetPoints(0x1A), data: 00 00 00
heatpump(0x20) -W-> boiler(0x08), ?(0x1E), data: 00 EF
heatpump(0x20) -B-> all(0x00), ?(0x02D7), data: 01 00 00 00 EF 19 02
```

Please test https://github.com/MichaelDvP/EMS-ESP32/releases/tag/latest
( use [EMS-ESP-3_7_0-dev_48-ESP32-16MB+.bin](https://github.com/MichaelDvP/EMS-ESP32/releases/download/latest/EMS-ESP-3_7_0-dev_48-ESP32-16MB+.bin) for the E32V2 board)
This should show MH210 as mixer and requests the mixer setting telegrams. 
Check, and please send updated system info and log to see the changes.

For the discovery: There are a lot of changes in 3.7. Try to recreate the config: In mqtt settings disable discovery (leave mqtt on), wait 5 minutes and reenable discovery. This should clear and rebuild the emsesp-discovery-configs on mqtt broker. (Or disable mqtt, use mqtt-explorer to delete all emsesp configs from broker and restart mqtt).

**Comment by Domoberry** [comment]
On Domoticz support for mqtt-ad value templates: I'll drop a note on the Domoticz forum on a (sort-off) similar question I asked, hoping to initiate direct contact between experts addressing mqtt-ad support going forward. Obviously, I would like to support, but I'm afraid I would not be able to help on coding level. See [https://www.domoticz.com/forum/posting.php?mode=reply&t=42755#preview](url)

On a new 3.7.0dev version for de ems-esp gateway: Thank you! I intend to temporary install this on on my gateway. Yet, as this will create some havoc for the existing setup (I do not have a test environment or similar), I plan to use a different mqtt-ad discovery topic such that on Domoticz side a complete new set up Domoticz-devices will be created. Once installed and running I'll share the logs.

**Comment by Domoberry** [comment]
Today I installed the latest dev version. The version you mentioned in the post above had probably been superseded with a new one (my bad, should have tested earlier), so I used latest available one: EMS-ESP-3_7_1-dev_1-ESP32-16MB+.bin. (if this is not correct, please let me know)
After installing this on the gateway, I downloaded the requested log files (see attached).
I noticed:
- Gateway now correctly mentions MH210 device now mentioned, as mixer modules and as heat pump (see screenshot attached
- About 45 devices created in Domoticz
- In Domoticz mqtt-AD hardware / setup many entities listed as configurable, but no values (they had values in 3.6.5, yet preferably these should show up as devices in Domoticz. This seems related to how Domoticz supports 'number' mqtt-ad config messages) (screenshot)
- Quite some errors in the Domoticz log (file attached)

Gateway System log
[241031 V3.7.1dev system log.txt](https://github.com/user-attachments/files/17586511/241031.V3.7.1dev.system.log.txt)
Gateway System info
[241031 V3.7.1dev emsesp_system_info.json](https://github.com/user-attachments/files/17586526/241031.V3.7.1dev.emsesp_system_info.json)
Gateway Devices screenshot
[241031 V3 7 1dev - Devices and Entities](https://github.com/user-attachments/assets/d29ac19a-5bc5-4e2c-91a0-e28789cf269e)
Domoticz configurable devices screenshot
[241031 V3 7 1dev setup configuration no values](https://github.com/user-attachments/assets/26bf3a4e-5f3e-4959-a6ad-248b95a1aa42)
https://github.com/user-attachments/assets/62879bfc-6a19-41b4-81a4-596e208021b9)
Domoticz errors
[241031 V3.7.1.dev Domoticz erros.txt](https://github.com/user-attachments/files/17586585/241031.V3.7.1.dev.Domoticz.erros.txt)

**Reply by proddy** [reply]
I'll remove the templating conditions as Domoticz doesn't support that it seems. This will solve the "is defined else 0" errorsI see in your log file.

For the other errors you will need to help us as I don't use Domoticz or plan ever to install it. Take for example `Error: TST: Exception (GetValueFromTemplate): stoi! (Template: value_json.dhw[curflow])`. Look up `curflow` using MQTTExplorer and see what type it is (sensor, select, switch...) and whether it has a value.

**Comment by Domoberry** [comment]
Thanks, will install once available.
In the mean time, with the risk of asking a akward question: I do not seem to be able to downgrade to 3.5.6 using the default process via the upload option in the webui. Should I better follow different steps to go back to v3.5.6?

**Reply by proddy** [reply]
never tried, but uploading a 3.6.5 binary should work. perhaps it ended up in a different partition? It's hard to tell from here.

**Comment by Domoberry** [comment]
quick feedback on v3.7.1-dev1:
Quite an improvement for Domoticz.
- many more devices in Domoticz (due to number -> sensor change I would guess?)
- less errors in the Domoticz log (see file)
added gateway log files FYI.
[241031 V3.7.1-dev1 emsesp_system_info.json](https://github.com/user-attachments/files/17587710/241031.V3.7.1-dev1.emsesp_system_info.json)
[241031 V3.7.1-dev1 system log.txt](https://github.com/user-attachments/files/17587711/241031.V3.7.1-dev1.system.log.txt)
[241031 V3.7.1-dev1 Domoticz errors.txt](https://github.com/user-attachments/files/17587712/241031.V3.7.1-dev1.Domoticz.errors.txt)

**Reply by proddy** [reply]
Thanks for checking! There are two settings for Domoticz - the one with (latest) uses Numbers as this was confirmed last year by another user. There's an EMS-ESP issue for it somewhere. Documented here: https://emsesp.org/Configuring/#mqtt-setup

I'll look into those other errors, but still flying blind.

**Reply by proddy** [reply]
maybe domoticz can't handle nested objects like `value_json.hc1[seltemp])`. Perhaps edit that in MQTTExplorer to  `value_json[hc1].seltemp` or something similar to see if that works?

**Reply by MichaelDvP** [reply]
Looks like domoticz does not like https://github.com/emsesp/EMS-ESP32/blob/13dfc87c57c1cbb7020ed58d8c24a793a5b749f6/src/mqtt.cpp#L1105
We can try to use `"value_json['%s']['%s']"`  or with maybe some errors for special characters `"value_json.%s.%s"`
Will HA accept `"value_json['%s']['%s']"` or should we only adapt for domoticz?

**Reply by proddy** [reply]
We need someone to modify the config topics locally and test, so we know which format works. I'll do a test as well in HA.

**Reply by proddy** [reply]
Quick test in HA (Developer Tools->Template) and those formats also work
```yaml
{% 
set value_json = {"reset":"","heatingoff":"off","heatingactive":"off","tapwateractive":"off","selflowtemp":5,"heatingpumpmod":0,"curflowtemp":49.7,"rettemp":45.1,"syspress":1.6,"boiltemp":53.1,"burngas":"off","burngas2":"off","flamecurr":0.0,"fanwork":"off","ignwork":"off","oilpreheat":"off","burnminpower":0,"burnmaxpower":50,"burnminperiod":10,"boilhyston":-6,"boilhystoff":6,"boil2hyston":0,"boil2hystoff":0,"curveon":"off","curvebase":20,"curveend":90,"summertemp":16,"nofrostmode":"off","nofrosttemp":5,"heatingactivated":"on","heatingtemp":70,"heatingpump":"off","pumpmodmax":70,"pumpmodmin":50,"pumpmode":"proportional","pumpdelay":2,"pumpontemp":0,"selburnpow":0,"curburnpow":0,"burnstarts":426945,"burnworkmin":746599,"burn2workmin":0,"heatworkmin":614654,"heatstarts":49673,"ubauptime":6120638,"lastcode":"6A(227) 11.07.2024 09:24 (0 min)","servicecode":"0H","servicecodenumber":203,"maintenancemessage":"H00","maintenance":"manual","maintenancetime":6000,"maintenancedate":"01.01.2012","nompower":30,"nrgtotal":892.90,"nrgheat":527.87,"dhw":{"tapactivated":"on","settemp":62,"seltemp":60,"type":"flow","comfort":"eco","flowtempoffset":40,"maxpower":100,"circpump":"off","chargetype":"3-way valve","hyston":-5,"hystoff":0,"disinfectiontemp":70,"circmode":"off","circ":"off","curtemp":39.8,"curflow":0.0,"storagetemp1":39.8,"activated":"on","onetime":"off","disinfecting":"off","charging":"off","recharging":"off","tempok":"on","active":"off","3wayvalve":"on","starts":377272,"workm":131945,"nrg":365.02}}
%}

{{ value_json.dhw.curtemp }}
{{ value_json['dhw'].curtemp }}
{{ value_json['dhw']['curtemp']}}
```

all show `39.8`

HA uses [Jinja2](https://jinja.palletsprojects.com/en/latest/templates/). No idea what Domoticz uses. Probably something from the late '90s ;-) Just kidding...

**Reply by MichaelDvP** [reply]
Then change to `"value_json['%s']['%s']"`, it's better than any mixed format and allows special characters in name and tag. 
If domoticz still complains, we need a special handling for domoticz.

**Reply by proddy** [reply]
I'll add to 3.7.1-dev-2 for testing

**Comment by Domoberry** [comment]
Just did a test drive with 3.7.1-dev.3, and apart from thanking you for the work done, here are the log files and a brief feedback.
Note: I'm using discovery type 'domoticz (latest)' (will try 'domoticz' some time later)

Short summary:
- Screenshot on detected EMS devices and entities 
   [241101 V3 7 1-dev3 EMS devices and entities](https://github.com/user-attachments/assets/5e925868-9f62-47b1-a3ef-daf85b14a1fb)
- EMS-ESP system info
  [241101 V3.7.1-dev3 emsesp_system_info.json](https://github.com/user-attachments/files/17600937/241101.V3.7.1-dev3.emsesp_system_info.json)
- EMS-ESP log
  [241101 V3.7.1-dev3 system log.txt](https://github.com/user-attachments/files/17600944/241101.V3.7.1-dev3.system.log.txt)   
- 35 entities as Domoticz lines in the 'mqtt-ad hw setup', which can be modified again
- 116 entities as normal Domoticz devices
- Domoticz log only shows two errors at the moment.
[241101 Domoticz Errors.txt](https://github.com/user-attachments/files/17600928/241101.Domoticz.Errors.txt)

On the domoticz errors: if time permits, I will try to dig up the related EMS-ESP mqtt-ad config messages and see why it might give an error when processing the JSON message with all values.

... impressed by the progress and the speed it is done!

**Reply by proddy** [reply]
I'm impressed in how thorough and detailed your responses are! You only really need to post the Domoticz logs :-) I think those two remaining errors are left over in your MQTT server as retained messages. I can't see how EMS-ESP generates them in the code. Can you check again and perhaps manually remove them?

**Comment by Brugje** [comment]
I have dev-3 installed.All the sensors and switches are in domoticz except those who are changeble. They are not recognized in domoticz.
I have highlited those who I would like to see in domoticz.
![image](https://github.com/user-attachments/assets/6d1a23d7-d98e-4c84-8059-1604074391d2)
In mqtt explorer these vallues are there.  under select and switch are in domoticz. All beneath number are missing
![image](https://github.com/user-attachments/assets/ccab6e7d-3960-435f-89f0-e0b40a29d103)
Is this something for here or for the domoticz forum.

**Reply by proddy** [reply]
I don't know if number, select and switch are supported in Domoticz MQTT discovery. If you could find out, or point me to some documentation that would be helpful

**Reply by MichaelDvP** [reply]
Afair we have faced the number issue with domoticz earlier and DiscoveryType:Domoticz set this to sensor, but DiscoveryType Domoticz-latest set's number again.

**Reply by Brugje** [reply]
I found this discussion on the domoticz forum. Started by this topic starter domoberry a few weeks ago. [https://www.domoticz.com/forum/viewtopic.php?t=42755](url)

**Reply by proddy** [reply]
We learned in the latest version of Domoticz they support Number. I think we just need to find a default uom for those entities (from the system like RxFails) and then we're complaint again

**Reply by proddy** [reply]
Let's continue with the Domoticz discussion in this issue https://github.com/emsesp/EMS-ESP32/issues/2177

**Comment by proddy** [comment]
I'll close this discussion, as I believe we've covered the support and questions related to the MH210

**Reply by Domoberry** [reply, marked as answer]
you did, thanks!
